File: app/src/main/java/app/drewromanyk/com/minesweeper/activities/GameActivity.java
package app.drewromanyk.com.minesweeper.activities;

import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.ActivityInfo;
import android.os.Vibrator;
import android.support.v4.app.NavUtils;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.Toolbar;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.Surface;
import android.view.View;
import android.view.WindowManager;
import android.widget.HorizontalScrollView;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.android.gms.analytics.HitBuilders;

import app.drewromanyk.com.minesweeper.enums.GameSoundType;
import app.drewromanyk.com.minesweeper.enums.UiThemeModeEnum;
import app.drewromanyk.com.minesweeper.models.Board;
import app.drewromanyk.com.minesweeper.R;
import app.drewromanyk.com.minesweeper.enums.GameDifficulty;
import app.drewromanyk.com.minesweeper.enums.GameStatus;
import app.drewromanyk.com.minesweeper.enums.ResultCodes;
import app.drewromanyk.com.minesweeper.models.YesNoDialogInfo;
import app.drewromanyk.com.minesweeper.util.DialogInfoUtils;
import app.drewromanyk.com.minesweeper.util.Helper;
import app.drewromanyk.com.minesweeper.util.SoundPlayer;
import app.drewromanyk.com.minesweeper.util.UserPrefStorage;
import app.drewromanyk.com.minesweeper.views.BoardInfoView;


public class GameActivity extends BackActivity {

    private static final String TAG = "GameActivity";

    private Board minesweeperBoard;
    // SOUNDS
    private SoundPlayer sound_player;
    // UI ELEMENTS
    private ScrollView vScroll;
    private View boardBackground;
    public BoardInfoView boardInfoView;
    public MenuItem refreshButton;
    public MenuItem flagButton;
    //SETTING
    private boolean flagMode = false;
    public boolean gamePlaying = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game);
        setupAds();

        setupToolbar((Toolbar) findViewById(R.id.toolbar), getString(R.string.nav_play));

        setupBoardInfoLayout();
        setupBoardLayout(savedInstanceState);

        applySettings();
    }

    @Override
    protected void setupToolbar(Toolbar toolbar, String title) {
        super.setupToolbar(toolbar, title);

        toolbar.inflateMenu(R.menu.menu_game);
        flagButton = toolbar.getMenu().findItem(R.id.action_flag);
        refreshButton = toolbar.getMenu().findItem(R.id.action_refresh);
    }

    private void setupBoardInfoLayout() {
        boardInfoView = new BoardInfoView(
                (TextView) findViewById(R.id.timeKeeper),
                (TextView) findViewById(R.id.mineKeeper),
                (TextView) findViewById(R.id.scoreKeeper));
    }

    private void setupBoardLayout(Bundle savedInstanceState) {
        boardBackground = findViewById(R.id.boardBackground);
        boardBackground.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (minesweeperBoard != null && minesweeperBoard.getGameStatus() == GameStatus.PLAYING && UserPrefStorage.getSwiftChange(v.getContext())) {
                    changeFlagMode(minesweeperBoard);
                }
            }
        });

        setupBiDirectionalScrolling();
        setupBoard(savedInstanceState == null);
    }

    private void applySettings() {
        if (UserPrefStorage.getScreenOn(this)) {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }

        if (UserPrefStorage.getLockRotate(this)) {
            int orientation;
            int rotation = ((WindowManager) getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay().getRotation();
            switch (rotation) {
                case Surface.ROTATION_0:
                    orientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT;
                    break;
                case Surface.ROTATION_90:
                    orientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE;
                    break;
                case Surface.ROTATION_180:
                    orientation = ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT;
                    break;
                default:
                case Surface.ROTATION_270:
                    orientation = ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE;
                    break;
            }

            setRequestedOrientation(orientation);
        }

        // Theme Changing Background
        UiThemeModeEnum uiThemeMode = UserPrefStorage.getUiThemeMode(this);
        switch (uiThemeMode) {
            case LIGHT:
                boardBackground.setBackgroundColor(ContextCompat.getColor(this, R.color.light_background));
                break;
            case DARK:
                boardBackground.setBackgroundColor(ContextCompat.getColor(this, R.color.dark_background));
                break;
            case AMOLED:
                boardBackground.setBackgroundColor(ContextCompat.getColor(this, R.color.amoled_background));
                break;
            case CLASSICAL:
                boardBackground.setBackgroundColor(ContextCompat.getColor(this, R.color.light_background));
                break;
        }
        minesweeperBoard.updateCellSize();
    }

    @Override
    protected void onPause() {
        super.onPause();
        gamePlaying = (minesweeperBoard.getGameStatus() == GameStatus.PLAYING);
        if (minesweeperBoard != null) {
            minesweeperBoard.stopGameTime();
        }

        UserPrefStorage.saveBoardInfo(this, minesweeperBoard);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (minesweeperBoard != null && !minesweeperBoard.getFirstRound() && gamePlaying)
            minesweeperBoard.startGameTime();

        Helper.screenViewOnGoogleAnalytics(this, "Game");
    }

    @Override
    protected void onStart() {
        super.onStart();
        sound_player = new SoundPlayer(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        sound_player.release();
        sound_player = null;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_game, menu);
        flagButton = menu.findItem(R.id.action_flag);
        refreshButton = menu.findItem(R.id.action_refresh);

        View action_zoomin = findViewById(R.id.action_zoomin);
        if (action_zoomin != null) action_zoomin.setOnLongClickListener(
                new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        if (minesweeperBoard != null) {
                            minesweeperBoard.zoomInFully();
                        }
                        return true;
                    }
                }
        );
        View action_zoomout = findViewById(R.id.action_zoomout);
        if (action_zoomout != null) action_zoomout.setOnLongClickListener(
                new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        if (minesweeperBoard != null) {
                            minesweeperBoard.zoomOutFully();
                        }
                        return true;
                    }
                }
        );

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_refresh:
                if (minesweeperBoard.getGameStatus() == GameStatus.PLAYING) {
                    YesNoDialogInfo dialogInfo = DialogInfoUtils.getInstance(this).getDialogInfo(ResultCodes.RESTART_DIALOG.ordinal());
                    AlertDialog dialog = new AlertDialog.Builder(this)
                            .setTitle(dialogInfo.getTitle())
                            .setMessage(dialogInfo.getDescription())
                            .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    minesweeperBoard.gameOverByRestart();
                                    setupBoard(true);
                                }
                            })
                            .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                }
                            })
                            .create();
                    dialog.show();
                } else {
                    setupBoard(true);
                }
                return true;
            case R.id.action_flag:
                changeFlagMode(minesweeperBoard);
                return true;
            case R.id.action_zoomin:
                if (minesweeperBoard != null) {
                    minesweeperBoard.zoomIn();
                }
                return true;
            case R.id.action_zoomout:
                if (minesweeperBoard != null) {
                    minesweeperBoard.zoomOut();
                }
                return true;
            case android.R.id.home:
                NavUtils.navigateUpFromSameTask(this);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        int action = event.getAction();
        int keyCode = event.getKeyCode();
        if (!isFinishing()) {
            if (UserPrefStorage.getVolumeButton(this)) {
                switch (keyCode) {
                    case KeyEvent.KEYCODE_VOLUME_UP:
                        if (action == KeyEvent.ACTION_DOWN) {
                            changeFlagMode(minesweeperBoard);
                        }
                        return true;
                    case KeyEvent.KEYCODE_VOLUME_DOWN:
                        if (action == KeyEvent.ACTION_DOWN) {
                            changeFlagMode(minesweeperBoard);
                        }
                        return true;
                }
            }
        }
        return super.dispatchKeyEvent(event);
    }

    /*
     * USER FEEDBACK
     */

    public void playSoundEffects(GameSoundType type) {
        sound_player.play(type);
    }

    public void vibrate() {
        if (UserPrefStorage.getVibrate(this)) {
            Vibrator v = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
            v.vibrate(100);
        }
    }

    /*
     * FLAG MODE
     */

    public boolean getFlagMode() {
        return flagMode;
    }

    public void changeFlagMode(Board board) {
        flagMode = !flagMode;
        int icon = (flagMode) ? R.drawable.ic_action_flag : R.drawable.ic_action_notflag;
        flagButton.setIcon(icon);
        board.updateCellSize();
    }

    /*
     * SETUP BOARD
     */

    private void setupBoard(boolean savedStateIsEmpty) {
        GameDifficulty gameDifficulty = GameDifficulty.valueOf(getIntent().getStringExtra("gameDifficulty"));
        if (gamePlaying) {
            gameDifficulty = minesweeperBoard.getGameDifficulty();
        }
        if (minesweeperBoard != null) {
            minesweeperBoard.stopGameTime();
        }

        // Create or load a new board
        if ((gameDifficulty == GameDifficulty.RESUME && minesweeperBoard == null) || !savedStateIsEmpty) {
            minesweeperBoard = UserPrefStorage.loadSavedBoard(this, false);
        } else {
            // Set to custom game defaults
            int columns = UserPrefStorage.getColumnCount(this);
            int rows = UserPrefStorage.getRowCount(this);
            int mineCount = UserPrefStorage.getMineCount(this);

            if (gameDifficulty == GameDifficulty.EASY) {
                columns = 9;
                rows = 9;
                mineCount = 10;
            } else if (gameDifficulty == GameDifficulty.MEDIUM) {
                columns = 16;
                rows = 16;
                mineCount = 40;
            } else if (gameDifficulty == GameDifficulty.EXPERT) {
                columns = 30;
                rows = 16;
                mineCount = 99;
            }

            boardInfoView.resetInfo(mineCount);
            double gameCellScale = (minesweeperBoard != null) ? minesweeperBoard.getGameCellScale() : 1;
            minesweeperBoard = new Board(rows, columns, mineCount, gameCellScale, gameDifficulty, this);
        }

        assert minesweeperBoard != null;
        assert getSupportActionBar() != null;

        switch (minesweeperBoard.getGameDifficulty()) {
            case EASY:
                getSupportActionBar().setTitle(getString(R.string.game_difficulty_easy));
                break;
            case MEDIUM:
                getSupportActionBar().setTitle(getString(R.string.game_difficulty_medium));
                break;
            case EXPERT:
                getSupportActionBar().setTitle(getString(R.string.game_difficulty_expert));
                break;
            default:
                getSupportActionBar().setTitle(getString(R.string.game_difficulty_custom));
                break;
        }

        gamePlaying = (minesweeperBoard.getGameStatus() == GameStatus.PLAYING);
        addBoardToLayout();
    }

    private void setupBiDirectionalScrolling() {

        final HorizontalScrollView hScroll = (HorizontalScrollView) findViewById(R.id.scrollHorizontal);
        if (hScroll == null) return;

        vScroll = (ScrollView) findViewById(R.id.scrollVertical);

        vScroll.setOnTouchListener(new View.OnTouchListener() { //inner scroll listener
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                return false;
            }
        });
        hScroll.setOnTouchListener(new View.OnTouchListener() { //outer scroll listener
            private float mx, my, curX, curY;
            private boolean started = false;

            @Override
            public boolean onTouch(View v, MotionEvent event) {
                curX = event.getX();
                curY = event.getY();
                int dx = (int) (mx - curX);
                int dy = (int) (my - curY);
                switch (event.getAction()) {
                    case MotionEvent.ACTION_MOVE:
                        if (started) {
                            vScroll.scrollBy(0, dy);
                            hScroll.scrollBy(dx, 0);
                        } else {
                            started = true;
                        }
                        mx = curX;
                        my = curY;
                        break;
                    case MotionEvent.ACTION_UP:
                        vScroll.scrollBy(0, dy);
                        hScroll.scrollBy(dx, 0);
                        started = false;
                        break;
                }
                return true;
            }
        });
    }

    private void addBoardToLayout() {
        vScroll.removeAllViews();
        vScroll.addView(minesweeperBoard.getLayout());
        if (flagMode) {
            changeFlagMode(minesweeperBoard);
        }
    }
}

File: app/src/main/kotlin/app/drewromanyk/com/minesweeper/activities/GameActivity.kt
package app.drewromanyk.com.minesweeper.activities

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.view.Surface
import android.view.View
import android.view.WindowManager
import android.widget.HorizontalScrollView
import android.widget.ScrollView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.core.app.NavUtils
import androidx.core.content.ContextCompat

import app.drewromanyk.com.minesweeper.R
import app.drewromanyk.com.minesweeper.enums.*
import app.drewromanyk.com.minesweeper.interfaces.MinesweeperUiHandler
import app.drewromanyk.com.minesweeper.util.DialogInfoUtils
import app.drewromanyk.com.minesweeper.util.Helper
import app.drewromanyk.com.minesweeper.util.UserPrefStorage
import app.drewromanyk.com.minesweeper.views.BoardInfoView
import app.drewromanyk.com.minesweeper.views.MinesweeperUI

/**
 * Created by Drew Romanyk on 5/19/17.
 * Activity that handles the whole game and its interactions
 */


class GameActivity : BackActivity() {

    private lateinit var minesweeperUI: MinesweeperUI
    // UI ELEMENTS
    private lateinit var vScroll: ScrollView
    private lateinit var boardBackground: View
    private lateinit var boardInfoView: BoardInfoView
    private lateinit var refreshButton: MenuItem
    private lateinit var flagButton: MenuItem

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)
        setupAds()

        setupToolbar(findViewById(R.id.toolbar), getString(R.string.nav_play))

        boardInfoView = BoardInfoView(
                findViewById(R.id.timeKeeper),
                findViewById(R.id.mineKeeper),
                findViewById(R.id.scoreKeeper))
        setupBoardLayout(savedInstanceState)

        applySettings()
    }

    override fun setupToolbar(toolbar: Toolbar, title: String) {
        super.setupToolbar(toolbar, title)

        toolbar.inflateMenu(R.menu.menu_game)
        flagButton = toolbar.menu.findItem(R.id.action_flag)
        refreshButton = toolbar.menu.findItem(R.id.action_refresh)
    }

    private fun setupBoardLayout(savedInstanceState: Bundle?) {
        boardBackground = findViewById(R.id.boardBackground)
        boardBackground.setOnClickListener {
            if (UserPrefStorage.getSwiftChange(this)) {
                minesweeperUI.switchClickMode()
            }
        }

        setupBiDirectionalScrolling()
        setupGame(savedInstanceState == null)
    }

    private fun applySettings() {
        if (UserPrefStorage.getScreenOn(this)) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }

        if (UserPrefStorage.getLockRotate(this)) {
            val rotation = (getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.rotation
            requestedOrientation = when (rotation) {
                Surface.ROTATION_0 -> ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
                Surface.ROTATION_90 -> ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
                Surface.ROTATION_180 -> ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT
                Surface.ROTATION_270 -> ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE
                else -> ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE
            }
        }

        // Theme Changing Background
        val uiThemeMode = UserPrefStorage.getUiThemeMode(this)
        boardBackground.setBackgroundColor(ContextCompat.getColor(this, uiThemeMode.color))
    }

    override fun onPause() {
        super.onPause()
        minesweeperUI.pauseTimer()
        minesweeperUI.save(this)
    }

    public override fun onResume() {
        super.onResume()
        minesweeperUI.resumeTimer()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_game, menu)
        flagButton = menu.findItem(R.id.action_flag)
        refreshButton = menu.findItem(R.id.action_refresh)

        val actionZoomIn = findViewById<View>(R.id.action_zoomin)
        actionZoomIn.setOnLongClickListener {
            minesweeperUI.zoomIn(true)
            true
        }
        val actionZoomOut = findViewById<View>(R.id.action_zoomout)
        actionZoomOut.setOnLongClickListener {
            minesweeperUI.zoomOut(true)
            true
        }

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.action_refresh -> {
                if (minesweeperUI.isPlaying()) {
                    val dialogInfo = DialogInfoUtils.getInstance(this).getDialogInfo(ResultCodes.RESTART_DIALOG.ordinal)
                    val dialog = AlertDialog.Builder(this)
                            .setTitle(dialogInfo.title)
                            .setMessage(dialogInfo.description)
                            .setPositiveButton(android.R.string.yes) { _, _ ->
                                minesweeperUI.reset(this)
                            }
                            .setNegativeButton(android.R.string.no) { _, _ -> }
                            .create()
                    dialog.show()
                } else {
                    minesweeperUI.reset(this)
                }
                return true
            }
            R.id.action_flag -> {
                minesweeperUI.switchClickMode()
                return true
            }
            R.id.action_zoomin -> {
                minesweeperUI.zoomIn(false)
                return true
            }
            R.id.action_zoomout -> {
                minesweeperUI.zoomOut(false)
                return true
            }
            android.R.id.home -> {
                NavUtils.navigateUpFromSameTask(this)
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        val action = event.action
        val keyCode = event.keyCode
        if (!isFinishing) {
            if (UserPrefStorage.getVolumeButton(this)) {
                when (keyCode) {
                    KeyEvent.KEYCODE_VOLUME_UP -> {
                        if (action == KeyEvent.ACTION_DOWN) {
                            minesweeperUI.switchClickMode()
                        }
                        return true
                    }
                    KeyEvent.KEYCODE_VOLUME_DOWN -> {
                        if (action == KeyEvent.ACTION_DOWN) {
                            minesweeperUI.switchClickMode()
                        }
                        return true
                    }
                }
            }
        }
        return super.dispatchKeyEvent(event)
    }

    /*
     * SETUP BOARD
     */

    private fun setupGame(savedStateIsEmpty: Boolean) {
        val gameDifficulty = GameDifficulty.valueOf(intent.getStringExtra("gameDifficulty"))

        minesweeperUI = MinesweeperUI(!savedStateIsEmpty, gameDifficulty, boardInfoView, this, object : MinesweeperUiHandler {
            override fun onGameTimerTick(gameTime: Long, score: Double) {
                runOnUiThread {
                    boardInfoView.setTimeKeeperText(gameTime)
                    boardInfoView.setScoreKeeperText(score)
                }
            }

            override fun onFlagChange(clickMode: ClickMode) {
                val icon = if (clickMode == ClickMode.FLAG)
                    R.drawable.ic_action_flag else R.drawable.ic_action_notflag
                flagButton.setIcon(icon)
            }

            override fun onVictory(score: Long, time: Long) {
                updateLeaderboards(GameStatus.VICTORY, minesweeperUI.gameDifficulty, score, time)
            }
        })

        supportActionBar?.title = minesweeperUI.gameDifficulty.getName(this)

        vScroll.removeAllViews()
        vScroll.addView(minesweeperUI.layout)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setupBiDirectionalScrolling() {

        val hScroll = findViewById<HorizontalScrollView>(R.id.scrollHorizontal)

        vScroll = findViewById(R.id.scrollVertical)

        //inner scroll listener
        vScroll.setOnTouchListener { _, _ ->
            false
        }

        //outer scroll listener
        hScroll.setOnTouchListener(object : View.OnTouchListener {
            private var mx: Float = 0.toFloat()
            private var my: Float = 0.toFloat()
            private var curX: Float = 0.toFloat()
            private var curY: Float = 0.toFloat()
            private var started = false

            override fun onTouch(v: View, event: MotionEvent): Boolean {
                curX = event.x
                curY = event.y
                val dx = (mx - curX).toInt()
                val dy = (my - curY).toInt()
                when (event.action) {
                    MotionEvent.ACTION_MOVE -> {
                        if (started) {
                            vScroll.scrollBy(0, dy)
                            hScroll.scrollBy(dx, 0)
                        } else {
                            started = true
                        }
                        mx = curX
                        my = curY
                    }
                    MotionEvent.ACTION_UP -> {
                        vScroll.scrollBy(0, dy)
                        hScroll.scrollBy(dx, 0)
                        started = false
                    }
                }
                return true
            }
        })
    }
}